packages/compass-query-bar/src/stores/query-bar-reducer.ts

```diff
@@ -294,13 +231,188 @@ export const openExportToLanguage = (): QueryBarThunkAction<void> => {
+  };
+};
+
+export const saveRecentAsFavorite = (
+  recentQuery: RecentQuery,
+  name: string
+): QueryBarThunkAction<Promise<boolean>> => {
+  return async (dispatch, getState, { favoriteQueryStorage }) => {
+    try {
+      const now = new Date();
+      const { _id, _host, _lastExecuted, _ns, ...baseQuery } = recentQuery;
+      const favoriteQuery: FavoriteQuery = {
+        ...getQueryAttributes(baseQuery),
+        _id,
+        _ns,
+        _lastExecuted,
+        _host: _host ?? getState().host,
+        _name: name,
+        _dateSaved: now,
+        _dateModified: now,
+      };
+
+      // add it in the favorite
+      await favoriteQueryStorage.updateAttributes(
+        favoriteQuery._id,
+        favoriteQuery
+      );
+
+      // update favorites
+      void dispatch(fetchFavorites());
+
+      // remove from recents
+      void dispatch(deleteRecentQuery(_id));
+
+      return true;
+    } catch (e) {
+      debug('Failed to save recent query as favorite', e);
+      return false;
+    }
+  };
+};
+
+export const deleteRecentQuery = (
+  id: string
+): QueryBarThunkAction<Promise<void>> => {
+  return async (dispatch, _getState, { recentQueryStorage }) => {
+    try {
+      await recentQueryStorage.delete(id);
+      return dispatch(fetchRecents());
+    } catch (e) {
+      debug('Failed to delete recent query', e);
+    }
+  };
+};
+
+export const deleteFavoriteQuery = (
+  id: string
+): QueryBarThunkAction<Promise<void>> => {
+  return async (dispatch, _getState, { favoriteQueryStorage }) => {
+    try {
+      await favoriteQueryStorage.delete(id);
+      return dispatch(fetchFavorites());
+    } catch (e) {
+      debug('Failed to delete favorite query', e);
+    }
+  };
+};
+
+const saveRecentQuery = (
+  query: Omit<BaseQuery, 'maxTimeMS'>
+): QueryBarThunkAction<Promise<void>> => {
+  return async (_dispatch, getState, { recentQueryStorage }) => {
+    try {
+      const { recentQueries, host, namespace } = getState();
+
+      const queryAttributes = getQueryAttributes(query);
+      // Ignore empty or default queries
+      if (_.isEmpty(queryAttributes)) {
+        return;
+      }
+
+      // Ignore duplicate queries
+      const existingQuery = recentQueries.find((recentQuery) =>
+        _.isEqual(getQueryAttributes(recentQuery), queryAttributes)
+      );
+
+      if (existingQuery) {
+        // update the existing query's lastExecuted to move it to the top
+        const updateAttributes = {
+          _host: existingQuery._host ?? host,
+          _lastExecuted: new Date(),
+        };
+        await recentQueryStorage.updateAttributes(
+          existingQuery._id,
+          updateAttributes
+        );
+        return;
+      }
+
+      // Keep length of each recent list to TOTAL_RECENTS_ALLOWED
+      if (recentQueries.length >= TOTAL_RECENTS_ALLOWED) {
+        const lastRecent = recentQueries[recentQueries.length - 1];
+        await recentQueryStorage.delete(lastRecent._id);
+      }
```

Comment:
Don't have the strongest preference here, but maybe this is something worth making part of the QueryStorage? I'm thinking that if someone wants to use queryStorage for saving queries in some other place it would be hard to communicate that this is a limit that should also be applied and checked for there. What do you think?

Reply:
That's a good point. I'll create a follow up for this.